### DISCUSSION TITLE
PetPlantResponse의 dDay 필드에 대한 문제점

### DISCUSSION THREAD

**Discussion by Choi-JJunho** [woowacourse-teams/2023-pium#165] [질문있어요] [answered]
> 아래 내용은 백엔드 코드에 대한 내용이라 프론트 분들께서는 코멘트의 [프론트 분들을 위한 간단 요약](https://github.com/woowacourse-teams/2023-pium/discussions/165#discussioncomment-6600723)부분을 참고해주시면 감사하겠습니다 🙇‍♂️

## 사건의 발단

```java

@Getter
@Builder
@NoArgsConstructor(access = AccessLevel.PRIVATE)
@AllArgsConstructor(access = AccessLevel.PRIVATE)
public class PetPlantResponse {

    private Long id;
    private String nickname;
    private String imageUrl;
    private String location;
    private String flowerpot;
    private String light;
    private String wind;
    private Integer waterCycle;
    private Long daySince;
    private Long dDay;

    @DateTimeFormat(pattern = "yyyy-MM-dd")
    private LocalDate birthDate;

    @DateTimeFormat(pattern = "yyyy-MM-dd")
    private LocalDate lastWaterDate;

    @DateTimeFormat(pattern = "yyyy-MM-dd")
    private LocalDate nextWaterDate;

    private DictionaryPlantResponse dictionaryPlant;

    @Getter
    @NoArgsConstructor
    public static class DictionaryPlantResponse {

        private Long id;
        private String name;

        @Builder
        private DictionaryPlantResponse(Long id, String name) {
            this.id = id;
            this.name = name;
        }
    }
}
```

위 객체에서 lombok이 dDay 필드에 대한 Getter를 생성하면서 한가지 문제가 발생했습니다.

<img width="160" alt="image" src="https://github.com/woowacourse-teams/2023-pium/assets/49794401/8e0a60cf-a5ec-433f-985b-4d5f81832ddc">

lombok은 위와같이 `getDDay()` 와 같은 형태로 getter를 생성합니다.

객체를 json 형태로 직렬화하는 과정에서 jackson의 objectMapper는 getter를 참조하여 json을 구성합니다.

<img width="294" alt="image" src="https://github.com/woowacourse-teams/2023-pium/assets/49794401/fd0bfe7b-f125-4ae6-87b5-b342746e2ab1">

결과는 기대했던 형태의 `dDay`가 아닌 `dday`가 나오게 됩니다.

이 때문에 테스트를 수행하기 위해 객체를 역직렬화 하는 과정에서 json의 `dday` 를 Response 객체의 `dDay`로 파싱을 하지 못하는 상황이 일어납니다.

<img width="1394" alt="image" src="https://github.com/woowacourse-teams/2023-pium/assets/49794401/edec8f8e-2b07-4e10-b539-6c430bbfaa71">

## 제안

위와 같은 문제상황에서 다음과 같은 해결방안들이 존재합니다.

1. 필드명을 dDay가 아닌 다른 이름의 변수를 사용한다.
2. 필드명을 dday로 수정한다.
3. Lombok의 `@Getter`를 사용하지 않고 getter를 직접 재정의한다. (`getdDay()`)

> `@JsonProperty`를 사용하는 방법도 존재하나 이 경우 직렬화 과정에서 lombok의 getter까지 같이 읽어버리는 문제가 발생하여 해당 방식은 제외했습니다.
<img width="460" alt="image" src="https://github.com/woowacourse-teams/2023-pium/assets/49794401/d3cce74d-e1a1-4ba6-9f1c-1a971804b46c">
<img width="299" alt="image" src="https://github.com/woowacourse-teams/2023-pium/assets/49794401/1d05be13-4abc-4c3b-98a2-5622a580d532">

개인적으로는 가장 변경소요가 적은 `필드명을 dday로 수정한다.`를 희망하고있습니다.
이에 추가적으로 이뤄져야하는 작업은 다음과 같습니다.

- API 명세 수정

이에 대한 의견 부탁드립니다

**Comment by Kim0914** [comment]
두근두근

**Comment by Choi-JJunho** [comment]
## 프론트 분들을 위한 간단 요약

백엔드의 구현 및 테스트 과정에서 일부 이슈가 존재해 API 명세에서 필드명을 변경해야하는 상황이 생겼습니다.
API 명세에서 `dDay` -> `dday`로 변경하는 것에 대한 의견 부탁드립니다!

### 영향을 받는 API

- 리마인더 - 조회
```json
{
	"data": [
		{
			"petPlantId": 1,
			"image": "image.com",
			"nickName": "참새나무",
			"dictionaryPlantName": "알로카시아",
			"dDay": 3, //  음수면 할일 | 0이면 오늘 | 양수면 지각
			"nextWaterDate": "2023-07-07",
		},
		{
			"petPlantId": 2,
			"image": "image.com",
			"nickName": "그레이나무",
			"dictionaryPlantName": "스투키",
			"dDay": 3, //  음수면 할일 | 0이면 오늘 | 양수면 지각
			"nextWaterDate": "2023-07-16"
		}
	]
}
```

**Reply by hozzijeong** [reply]
뭔가 다른 변수명이 있을까요??

**Reply by hozzijeong** [reply]
저는 dday가 좋을것 같긴 합니다

**Comment by Kim0914** [comment]
`dday`로 통일하는 방법이 제일 좋아보이네요 !

크게 수정할 요소도 없고 `@JsonProperty`나 `lombok`과 같은 라이브러리에 의존을 받지 않을 수 있을 것 같네요 ~

**Comment by yeonkkk** [comment]
저도 필드명을 다른 이름으로 변경하거나, `dday`로 수정하는데 동의합니다.
`Lombok`의 `@Getter`를 사용하지 않고 `getter`를 직접 재정의하게 되면 통일성이 깨지는 느낌이 들 것 같아요.
그리고 주노 말대로 필드명을 변경하는 작업이 가장 변경 사항이 적어서 좋을 것 같아요 ~~

**Comment by WaiNaat** [comment, marked as answer]
dday로 가시죠

**Comment by Choi-JJunho** [comment]
dDay -> dday로 변경하는것으로 결론짓겠습니다!

추가로 `리마인더 - 조회` API 내용 수정완료했습니다.

```json
{
	"data": [
		{
			"petPlantId": 1,
			"image": "image.com",
			"nickName": "참새나무",
			"dictionaryPlantName": "알로카시아",
			"dday": 3, //  음수면 할일 | 0이면 오늘 | 양수면 지각
			"nextWaterDate": "2023-07-07",
		},
		{
			"petPlantId": 2,
			"image": "image.com",
			"nickName": "그레이나무",
			"dictionaryPlantName": "스투키",
			"dday": 3, //  음수면 할일 | 0이면 오늘 | 양수면 지각
			"nextWaterDate": "2023-07-16"
		}
	]
}
```
